### PR TITLE
Fix #66

### LINKED ISSUE #66: database-relation-broken hook fails with error
Charm goes into a blocked state when removing a consuming application or relation. The database-relation-broken hook gets the following traceback as it assumes that the data is available in the relation at this point (it isn't).

unit-mysql-k8s-0: 19:59:42 DEBUG unit.mysql-k8s/0.juju-log database:2: Using local storage: /var/lib/juju/agents/unit-mysql-k8s-0/charm/.unit-state.db already exists
unit-keystone-operator-0: 19:59:42 ERROR juju.worker.uniter pebble poll failed for container "keystone": failed to get pebble info: cannot obtain system details: cannot communicate with server: Get "http://localhost/v1/system-info": dial unix /charm/containers/keystone/pebble.socket: connect: no such file or directory
unit-mysql-k8s-0: 19:59:42 DEBUG unit.mysql-k8s/0.juju-log database:2: new_auth_plugin: caching_sha2_password
unit-mysql-k8s-0: 19:59:42 DEBUG unit.mysql-k8s/0.juju-log database:2: MySQL Provider is available
unit-mysql-k8s-0: 19:59:42 DEBUG unit.mysql-k8s/0.juju-log database:2: Emitting Juju event database_relation_broken.
unit-mysql-k8s-0: 19:59:42 ERROR unit.mysql-k8s/0.juju-log database:2: Uncaught exception while in charm code:
Traceback (most recent call last):
  File "./src/charm.py", line 284, in <module>
    main(MySQLCharm)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/ops/main.py", line 409, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/ops/main.py", line 143, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/ops/framework.py", line 278, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/ops/framework.py", line 722, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/ops/framework.py", line 769, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/src/mysqlprovider.py", line 89, in _on_database_relation_broken
    databases = json.loads(
  File "/usr/lib/python3.8/json/__init__.py", line 341, in loads
    raise TypeError(f'the JSON object must be str, bytes or bytearray, '
TypeError: the JSON object must be str, bytes or bytearray, not NoneType
unit-mysql-k8s-0: 19:59:42 ERROR juju.worker.uniter.operation hook "database-relation-broken" (via hook dispatching script: dispatch) failed: exit status 1
^C

### PR DESCRIPTION
**Pull request by Abuelodelanada**

### CODE REVIEW

**Comment by sed-i** [comment]
I don't understand from the diff how a NoneType is avoided here (based on #66 traceback):
https://github.com/canonical/mysql-operator/blob/cdd3c88307f66f4e4b718ac5858fdde2a988b91f/src/mysqlprovider.py#L88-L91

**Comment by Abuelodelanada** [comment]
> I don't understand from the diff how a NoneType is avoided here (based on #66 traceback):
> 
> https://github.com/canonical/mysql-operator/blob/cdd3c88307f66f4e4b718ac5858fdde2a988b91f/src/mysqlprovider.py#L88-L91

It's a sync issue, despite of the fact that the container is running , sometimes mysql isn't ready when getting the version, and this leads to a NoneType when removing the relation.
Since we know which mysql version we are using, I simplified the code. 
Anyway when I get back to work with this charm, I will refactor the code to make it easier.

File: src/mysqlserver.py [hunk 1]
```diff
@@ -7,9 +7,9 @@
 import string
 
 from mysql.connector import connect, Error
-from typing import Union
 
 logger = logging.getLogger(__name__)
+MYSQL_VERSION = "8.0.26"
```
Comment (sed-i) [inline src/mysqlserver.py:12]: will need to remember to keep this in sync with the uploaded resource (image)

**Review by sed-i — APPROVED** [review]

File: src/mysqlserver.py (hunk 1 above)
Comment (Abuelodelanada) [reply src/mysqlserver.py:12]: Sure!

**Comment by balbirthomas** [comment]
> Since we know which mysql version we are using, I simplified the code.
> Anyway when I get back to work actively with this charm, I will refactor the code to make it easier.

I am approving as a stop gap arrangement. You could defer events or use the `ProviderBase.ready()` to solve such problems. Hardcoding version is not a good idea. 

Also I would put an extensive commit message explaining what the bug was and why the change fixes the bug.

**Review by balbirthomas — APPROVED** [review]

**Comment by Abuelodelanada** [comment]
> I am approving as a stop gap arrangement. You could defer events or use the `ProviderBase.ready()` to solve such problems. Hardcoding version is not a good idea.
> 
> Also I would put an extensive commit message explaining what the bug was and why the change fixes the bug.

Sure, I completely agree with you. When I get back to work actively with this charm, I will refactor the code.

**Comment by wolsen** [comment]
> > I don't understand from the diff how a NoneType is avoided here (based on #66 traceback):
> > https://github.com/canonical/mysql-operator/blob/cdd3c88307f66f4e4b718ac5858fdde2a988b91f/src/mysqlprovider.py#L88-L91
> 
> It's a sync issue, despite of the fact that the container is running , sometimes mysql isn't ready when getting the version, and this leads to a NoneType when removing the relation.
> Since we know which mysql version we are using, I simplified the code.
> Anyway when I get back to work actively with this charm, I will refactor the code to make it easier.

Actually, it was not a sync issue. I had a bug in my code which did not request databases from the mysql-operator charm and then I and removed the relation. I think as @sed-i was alluding to, the problem is that this code assumes that the databases already exists and attempts to unmarshall the relation data under that assumption.

**Pull request merged**